lib/ABI/AddABIAwarenessToFunctionPrototypes.cpp

```diff
@@ -0,0 +1,210 @@
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, Kind);
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, A.Type());
+    }
+    for (const auto &V : llvm::cast<RawFT>(&Type)->ReturnValues()) {
+      auto Kind = model::Register::primitiveKind(V.Location());
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, Kind);
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, V.Type());
+    }
+
+    const auto &Stack = llvm::cast<RawFT>(&Type)->StackArgumentsType();
+    revng_assert(Stack.Qualifiers().empty());
+    const model::TypePath &Internal = Stack.UnqualifiedType();
+    if (Internal.isValid())
+      Result = Result && rc_recur needsVectorRegisterSupportImpl(VH, Internal);
+  } break;
+
+  default:
+    revng_abort("Unknown type.");
+  }
+
+  if (Result)
+    VH.setVerified(&Type);
+
+  VH.verificationCompleted(&Type);
+
+  rc_return VH.maybeFail(Result);
+}
+
+static bool needsVectorRegisterSupport(model::VerifyHelper &VH,
+                                       const model::TypePath &Type) {
+  return !needsVectorRegisterSupportImpl(VH, model::QualifiedType(Type, {}));
+}
+
+using namespace std::string_literals;
+
+class AddABIAwarenessToFunctionPrototypes {
+public:
+  static constexpr auto Name = "AddABIAwarenessToFunctionPrototypes";
+  inline static const std::tuple Options = { pipeline::Option("abi",
+                                                              "Invalid"s) };
+  std::vector<std::vector<pipeline::Kind *>> AcceptedKinds = {};
+
+  void run(pipeline::Context &Context, std::string InputABI) {
+    auto &Model = revng::getWritableModelFromContext(Context);
+
+    // Minimize the negative impact on binaries with ABI that is not fully
+    // supported by disabling the conversion by default.
+    //
+    // Use revng model pass on those functions to convert them manually.
+    //
+    // TODO: extend this list.
+    static constexpr std::array ABIsTheConversionIsEnabledFor = {
+      model::ABI::SystemV_x86_64,        model::ABI::SystemV_x86,
+      model::ABI::SystemV_x86_regparm_3, model::ABI::SystemV_x86_regparm_2,
+      model::ABI::SystemV_x86_regparm_1, model::ABI::AAPCS
+    };
+    if (!llvm::is_contained(ABIsTheConversionIsEnabledFor, Model->DefaultABI()))
+      return;
+
+    // This reuses the verification map within the `model::VerifyHelper` in
+    // an incompatible manner. DO NOT pass this object into a normal
+    // verification routine or things are going to break down.
+    model::VerifyHelper VectorVH;
+
+    // This is a normal `model::VerifyHelper`.
+    model::VerifyHelper ValidationVH;
+
+    std::optional<model::ABI::Values> ABI = model::ABI::fromName(InputABI);
+    if (ABI.value() == model::ABI::Values::Invalid)
+      ABI = std::nullopt;
+
+    // Choose the applicable functions and run the conversion with the default
+    // ABI (`Model->DefaultABI`) for them.
```

Comment:
We're not using `DefaultABI` in all cases.

Reply:
Reworded.